### PR TITLE
[Bundle 6/n] Prune g_chainman usage in auxiliary modules

### PR DESCRIPTION
**Pull request by dongcarl**
Overall PR: #20158 (tree-wide: De-globalize ChainstateManager)

The first 2 commits are fixups addressing review for the last bundle: #21391

NEW note:
1. I have opened #21766 which keeps track of potential improvements where the flaws already existed before the de-globalization work, please post on that issue about these improvements, thanks!

Note to reviewers:
1. This bundle may _apparently_ introduce usage of `g_chainman` or `::Chain(state|)Active()` globals, but these are resolved later on in the overall PR. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
2. There may be seemingly obvious local references to `ChainstateManager` or other validation objects which are not being used in callers of the current function in question, this is done intentionally to **_keep each commit centered around one function/method_** to ease review and to make the overall change systematic. We don't assume anything about our callers. Rest assured that once we are considering that particular caller in later commits, we will use the obvious local references. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
3. When changing a function/method that has many callers (e.g. `LookupBlockIndex` with 55 callers), it is sometimes easier (and less error-prone) to use a scripted-diff. When doing so, there will be 3 commits in sequence so that every commit compiles like so:
	1. Add `new_function`, make `old_function` a wrapper of `new_function`, divert all calls to `old_function` to `new_function` **in the local module only**
	2. Scripted-diff to divert all calls to `old_function` to `new_function` **in the rest of the codebase**
	3. Remove `old_function`

### CODE REVIEW

**Comment by laanwj** [comment]
Looks like this accumulated a silent merge conflict:
```
…/bitcoin/src/test/coinstatsindex_tests.cpp:35:28: error: too few arguments to function call, single argument 'active_chainstate' was not specified
    coin_stats_index.Start();
    ~~~~~~~~~~~~~~~~~~~~~~ ^
…/bitcoin/src/index/base.h:122:10: note: 'Start' declared here
    void Start(CChainState& active_chainstate);
         ^
1 error generated.
```

File: src/rest.cpp
Comment (ryanofsky) [inline src/rest.cpp:120]: In commit "rest: Add GetChainman function and use it" (a839078c20f4ac3d92e14af7b287f2f0fddbc61e)

It seems like this should say `if (!node_context || !node_context->chainman)` otherwise the http error won't be set in the null chainman case.

File: src/rest.cpp
Comment (ryanofsky) [inline src/rest.cpp:121]: In commit "rest: Add GetChainman function and use it" (a839078c20f4ac3d92e14af7b287f2f0fddbc61e)

If this condition is never expected, then maybe this should use `HTTP_INTERNAL_SERVER_ERROR` to flag that this is a bug in request handling, not an invalid request.

File: src/rest.cpp
Comment (ryanofsky) [inline src/rest.cpp:670]: In commit "rest: Add GetChainman function and use it" (a839078c20f4ac3d92e14af7b287f2f0fddbc61e)

re: "Someone can probably make this a bit prettier," your current version of this seems fine to me, but if you wanted to make it more minimal you could use CHECK_NONFATAL and catch the NonFatalCheckError exception in here, I think: https://github.com/bitcoin/bitcoin/blob/06d573f053c63eb6521c62d210c5924418ae2b83/src/rest.cpp#L699

File: src/miner.cpp
Comment (ryanofsky) [inline src/miner.cpp:42]: In commit "miner: Pass in chainman to RegenerateCommitments" (f5368a525b3d5b93d06c83101497de5aaa6fe1c8)

It seems like this could take blockman reference instead of a chainman reference to be a little more limited. Doesn't matter though, all version of this function seem about the same to me.

File: src/index/base.cpp
Comment (ryanofsky) [inline src/index/base.cpp:68]: In commit "index: Add chainstate member to BaseIndex" (0945b4117e0209d156a6b74c407f4d32e0cf2c1a)

Seems like active_chain could be used above too. Fine either way in case this was intentional, though.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 0945b4117e0209d156a6b74c407f4d32e0cf2c1a. Looks good. I like cleaning up the index code. Straightforward review, too.

**Comment by dongcarl** [comment]
Pushed 0945b4117e -> b04370dbb5
- Addressed: https://github.com/bitcoin/bitcoin/pull/21767#discussion_r627846989, https://github.com/bitcoin/bitcoin/pull/21767#discussion_r627847595, https://github.com/bitcoin/bitcoin/pull/21767#discussion_r627853739

File: src/miner.cpp
Comment (dongcarl) [reply src/miner.cpp:42]: Unfortunately if we take a blockman reference, then callers would need to take `cs_main` to reach into ChainMan for the blockman ref :-/

Something for another day I think! Added to: #21766

File: src/rest.cpp
Comment (ryanofsky) [reply src/rest.cpp:674]: re: https://github.com/bitcoin/bitcoin/pull/21767#discussion_r628449195

> Hmmmm, how could we differentiate between NOT_FOUND like the "no mempool"-case and INTERNAL_SERVER_ERROR like the "no Context/Chainman"-case here?

I didn't realize that was intentional and I do think the your approach is fine, but you wanted to use the NonFatalCheckError approach, you would just add:

```
class PageNotFoundError : public std::runtime_error
{
    using std::runtime_error::runtime_error;
};

...

throw PageNotFoundError{};

...

} catch (const PageNotFoundError&) {

```

**Review by ryanofsky — APPROVED** [review]
Code review ACK b04370dbb55e61e168cbf0aba3d29fb882a46bec. Just http error code change and new commit tweaking variable usage since previous review

File: src/rest.cpp
Comment (ryanofsky) [reply src/rest.cpp:674]: Oh I see see the mempool code you're referring to is preexisting. That would more incline towards following previous pattern like you're doing I think. I'm sure you could just set the 404 one way and the 500 the other way, too though. Again any approach seems fine and I was just responding to dissatisfaction in the commit message.

File: src/rest.cpp
Comment (jnewbery) [inline src/rest.cpp:125]: I don't think this is necessarily a bug. Once chainman is deglobalized, it should be theoretically possible to run without a chainman. I think just having a message "Chainman disabled or instance not found" (similar to if mempool isn't found) is sufficient.

File: src/miner.h
Comment (jnewbery) [inline src/miner.h:206]: Perhaps just pass in a `BlockManager&` here?

File: src/bench/duplicate_inputs.cpp
Comment (jnewbery) [inline src/bench/duplicate_inputs.cpp:28]: Does this need to be in the `cs_main` scope?

File: src/bench/duplicate_inputs.cpp
Comment (jnewbery) [inline src/bench/duplicate_inputs.cpp:29]:
```suggestion
    CBlockIndex* pindexPrev = testing_setup->m_node.chainman->ActiveTip();
```

File: src/index/base.cpp
Comment (jnewbery) [inline src/index/base.cpp:63]: I really don't like `CChain` becoming part of chainstate's public interface. It's too low level - code outside validation shouldn't be interacting directly with that object and should be using higher level interface functions.

File: src/index/base.h
Comment (jnewbery) [inline src/index/base.h:122]: Any reason that you decided to put this in the `Start()` method rather than the constructor. You have access to the `::ChainstateActive()` object at the point you construct the index objects.

Adding this parameter to the ctor would allow you to set `m_chainstate` in the initializer list and make it a const member.

File: src/miner.h
Comment (MarcoFalke) [reply src/miner.h:206]: This is the 4th time the same function gets changed around. I think we should decide what to settle on, and ideally provide a rationale for it. 

https://github.com/bitcoin/bitcoin/pull/21391#discussion_r615237925

File: src/rest.cpp
Comment (MarcoFalke) [reply src/rest.cpp:125]: What is the point of running a bitcoin full node without a blockchain?

Comment (jnewbery) [reply src/rest.cpp:125]: Testing mostly, but you could also imagine using Bitcoin Core as an addr relayer without a blockchain (eg http://www.erisian.com.au/bitcoin-core-dev/log-2020-12-16.html#l-583), or serving compact block indexes.

File: src/miner.h
Comment (jnewbery) [reply src/miner.h:206]: I think fundamentally the issue here is with `generateblock()`, which calls into miner.cpp to construct the block, then changes the transactions in that block, and then calls back into miner.cpp to fix up the coinbase commitment. It'd be a much simpler interface if `generateblock()` just passed in a const ref to the vector of transactions it wants to include in its first call to miner.cpp.

I'm going to mark this comment as resolved. The mining code is a bit of a mess, and can be cleaned up in other PRs.

**Comment by jamesob** [comment]
Concept ACK, full review soon

File: src/index/base.h
Comment (dongcarl) [reply src/index/base.h:122]: I believe I did this mostly because ctors are not inherited, whereas methods are. If I were to make `m_chainstate` be a const member and initialized in the ctor, I'd have to change the ctor of every subclass.

File: src/index/base.cpp
Comment (dongcarl) [reply src/index/base.cpp:63]: Do you mean the `m_chain` member specifically or any kind of `CChain`? It seems that we were already using a `CChain` previously when we were caling `::ChainActive()`. Happy to document and add to #21766 so nothing gets lost!

File: src/rest.cpp
Comment (dongcarl) [reply src/rest.cpp:125]: I've been making changes with the assumption that chainman is considered absolutely necessary. If we decide that running bitcoin without a chainman is a usecase we want to support, I think we can have a separate discussion about that. Added to: #21766

File: src/bench/duplicate_inputs.cpp
Comment (dongcarl) [reply src/bench/duplicate_inputs.cpp:28]: I don't think it necessarily has to be (at least not before more of AssumeUTXO is merged), in any case Bundle 7 will remove all of these assertions!

**Comment by dongcarl** [comment]
Pushed b04370dbb55e61e168cbf0aba3d29fb882a46bec -> 6ef9163cc4604177fbdab5b644f7b2af7a33204b
- Rebased over master

File: src/index/base.h
Comment (ariard) [inline src/index/base.h:80]: nit: Add comment "Provide access to the active chainstate, containing the current most-work chain". Maybe even as a member name `m_active_chainstate`

**Review by ariard — COMMENTED** [review]
utACK f9222a0c

File: src/index/base.cpp
Comment (fjahr) [inline src/index/base.cpp]: nit: could be `const`

**Review by fjahr — COMMENTED** [review]
light code review ACK 6ef9163cc4604177fbdab5b644f7b2af7a33204b

I did a thorough review of the changes but did not review all the previous PRs in this series so I could be missing something due to a lack of context, that's why only "light".

File: src/rest.cpp
Comment (dongcarl) [reply src/rest.cpp:117]: Would be happy to add to PR if you supply a simple enough diff!

**Comment by dongcarl** [comment]
Pushed 6ef9163cc4604177fbdab5b644f7b2af7a33204b -> 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a
- Added REST crash explanation to commit message per: https://github.com/bitcoin/bitcoin/pull/21767#discussion_r637169193

File: src/rest.cpp
Comment (dongcarl) [reply src/rest.cpp:117]: Added in commit message

File: src/index/base.h
Comment (dongcarl) [reply src/index/base.h:80]: Not sure `m_active_chainstate` makes too much sense here, as it's "the active chainstate at the time when `::Start` was called"

**Review by ryanofsky — APPROVED** [review]
Code review ACK 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a. Basically no change since last review except fixed rebase conflicts and a new comment about REST Ensure()

**Comment by jarolrod** [comment]
ACK  7a799c9

Easy to follow what's going on, the assert statements and tests also give assurance to there being no change in behavior. Nice to see how this gets wrapped up on the next bundle.

File: src/index/base.h
Comment (ariard) [reply src/index/base.h:80]: Sure, it's just with assumeutxo, we start to have multiple chainstate references through the codebase (`m_ibd_chainstate`, `m_snapshot_chainstate`, `m_m_active_chainstate`) and it could be confusing in the future. I don't think it changes anything for the indexes rn.

**Comment by ariard** [comment]
Code Review ACK 7a799c9

File: src/miner.cpp
Comment (jamesob) [inline src/miner.cpp:55]: Hm, shouldn't we be asserting here (instead of assigning) to preserve behavior?

File: src/index/base.h
Comment (jamesob) [reply src/index/base.h:80]: To @ariard's point, we will likely have to facilitate changing this pointer mid-execution when the user calls `ActivateSnapshot()`. I'll have to warm my caches on the indexing stuff a little more, but IIRC we want this class instance to have a reference to the current active chainstate (which obviously will be liable to change as assumeutxo progresses).

Comment (jamesob) [reply src/index/base.h:80]: That said, the reduction of global use is an obvious win, and it should be pretty easy to facilitate changing `m_chainstate` via some method that can be called by `ActivateSnapshot()`.

Comment (jamesob) [reply src/index/base.h:80]: Might even just make sense to pass in a `ChainstateManager` instance to keep with this class... but that can be done in a follow-up.

**Review by jamesob — COMMENTED** [review]
conditional ACK 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a ([`jamesob/ackr/21767.1.dongcarl.bundle_6_n_prune_g_chai`](https://github.com/jamesob/bitcoin/tree/ackr/21767.1.dongcarl.bundle_6_n_prune_g_chai))

Only hung up on the question about change in `miner.cpp` behavior. Everything else is informational. Good change!

- [x] fc1c28284 rpc/blockchain: Use existing blockman in gettxoutsetinfo
  
  Oneline fix.

- [x] 9ecade142 rest: Add GetChainman function and use it
 
  I don't know the REST code well but all changes seem consistent with 
  existing conventions.

- [x] e6b4aa6eb miner: Pass in chainman to RegenerateCommitments
  
  Question about assertion -> assignment change.

- [x] 91226eb91 bench: Use existing NodeContext in DuplicateInputs

  `ActiveTip()` nit as already noted by John.

- [x] f4a47a1fe bench: Use existing chainman in AssembleBlock
 
- [x] db33cde80 index: Add chainstate member to BaseIndex

  See questions around active chainstate management in the context
  of `ActivateSnapshot()` use. The indexer probably needs to be made
  aware of the new active chainstate, or it should just hold a reference
  to chainman and not have to worry about being notified.

- [x] 7a799c9c2 index: refactor-only: Reuse CChain ref

  Could probably be fixed-up into the previous commit.

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a ([`jamesob/ackr/21767.1.dongcarl.bundle_6_n_prune_g_chai`](https://github.com/jamesob/bitcoin/tree/ackr/21767.1.dongcarl.bundle_6_n_prune_g_chai))

-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmCxUQAACgkQepNdrbLE
TwViSw/+IpN1mVXXNgnNcNJ4dv+u9QjAqslyD7xq5IpJrDSI7dPIU8XF1ypOlV2I
AiUd4sOheAF/fkj+61f8uRvWO55DdFPhuwarAxwUna4YvL8E4ui1uplavDGrsgMa
mhZzx3PoDBm2RotKXF/zPdsbZ18mXeGPEbePiH1+r1+wbIYp/7Mj4z5Snp+OK3EE
tuWVf4efTYvymL1ZosI/4gIcc8FbpcxCMHVD5RJaN/pH5KickFzs7oDiaBNuzIB1
lgC837X5a4cjEN68LBXVGBVax2RkS+N09p162+6dK8mlcc9ob6JtBZxcW4B3YAwP
k8AIbNOqpK1SI5ayvaz/GLpHbdwup1tjZxabhzYIJYYx3feslcg2h45H444xBqn3
01YrnJeG5LxjSKEdYAjYNagE7zsRY2kykhMEodBj8kAU+whtmM5SIYAP40rHHe78
4vGqo0hN4ekvdoA+sE4paqccNJkHDlN0ben5l2phS3FVCG4fBxM2nZ3OduOklAWH
l8kZQSO2b2IwDn/2a7iTBVk+2oTZ9rO4MK1Qk3cBumSAPQQvF0O9ZckYcIO6agvU
XM0xghbsEpbT2UexxEoDWu2u0DbQYXu4fOBnX+s8XDEjowsoPFVylcKrz/6YVkrP
jpmZEoKL6VAfmAv7BlvOocjpJq/sNZgIQtV63a3Ekikew2K9Nvc=
=BpU9
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-4.19.0-16-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default --enable-fuzz --with-sanitizers=address,fuzzer,undefined CXX=/usr/local/bin/clang++ CC=/usr/local/bin/clang --disable-shared --with-pic --enable-benchmark=no --with-bignum=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --with-bignum=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/local/bin/clang++ -std=c++17 -mavx -mavx2 -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -fsanitize=address,fuzzer,undefined -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 11.1.0-++20210405104510+1fdec59bffc1-1~exp1~20210405085125.161
```

</p></details>

File: src/rest.cpp
Comment (fjahr) [reply src/rest.cpp:117]: Something like this maybe, if have to retouch: https://github.com/fjahr/bitcoin/commit/69187f82b1227cc33bc5733c742b8f8b9fcefea3

**Comment by fjahr** [comment]
re-ACK 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a

File: src/rest.cpp
Comment (MarcoFalke) [reply src/rest.cpp:117]: This creates a copy of chainman, no?

**Comment by MarcoFalke** [comment]
review ACK 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a 🌠

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 7a799c9c2b8652e780d1fd5e1bf7d05b026c1c1a 🌠
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgnkAv/eYSF8/6fmyM3n1fjSVmbmXT74brAbRsg0tMEuWb0rtk603x8E8QSAAot
PCSICg/oAsOXVIzdxOWNYpqhEiceILJNlQTjuNfswOu0cgJNNqCp/sYX3nWrlmaW
9dqsK0zMMLjnC/U2xzSzvvU6id++ipSQlyWkbAjt+n7UorVd2A8wCdQMrMCQejpz
pFB1K25WAH7KYiDIwLxy1ZWO/x094gQaVSk6DfIu+KhUSQTL03QunHCFjN3Fyoey
8drI7oGaL0GtA6mcZHjApXhJG4ts7kYI39WJ8zD9wIHotQ1e9earLar9L28WwCRE
fhdCBtjHGSTf1V6WKOZFkvYVbMyxQhyUVqdSwsZzlcNB7szkUJfmLZTMNSfKKYBS
MPuNUM5MCB9/atp2+YVEvCw3TdaXHxNJ1JDLvuFMAMOvCD5BuGrZklYKW2FhP4Ar
MsMO6AoCfzHJUVvdWlsxWYneMlN37Eg428rZOw9lUH9rdWavMMskwpcUe8MaYY5w
e3hP/g/H
=P/4c
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `e8caa0bbe435d430361ec68dd841345069e815f61d2cf90bf463dd41e3290516  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108e8caa0bbe435d430361ec68dd841345069e815f61d2cf90bf463dd41e3290516f010db371f291dcd7e4ee32989791a2f703408fff0102cd1cc3b289d479285b9b3314ffe257d08f10460b61788f0087d6c1e1ebfea06140083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0102e0740c1f9448775dc49479281075d9708f10460b61788f00837100840f65255910083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010a8588cd95ed21e4a51deb7538bc4aebb08f10460b61787f008e651d53631cd2fc20083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/miner.cpp
Comment (dongcarl) [reply src/miner.cpp:55]: So `::RegenerateCommitments` has been changed a few times now:
1. In 2c3ba00693, where we passed in a `BlockManager`, instead of reaching for `g_chainman.m_blockman`
    - Because we don't want to reach for globals anymore (the overall point of this PR series)
2. In cced0f46c9, where we passed in the specific `CBlockIndex` instead of a `BlockManager`
    - Because it is a more specific validation-related object than `BlockManager`
3. In e6b4aa6eb5, where we pass in a `ChainstateManger` instead of the specific `CBlockIndex`
    - Because then we don't have to worry about callers violating the invariant checked here: https://github.com/bitcoin/bitcoin/blob/cced0f46c9133e0fc6211e987421ad1d9be1a399/src/miner.cpp#L48 

Just to be clear, the original behaviour does not do any checking/assertions:
https://github.com/bitcoin/bitcoin/blob/2afcf24408b4453e4418ebfb326b141f6ea8647c/src/miner.cpp#L42-L51

So overall, I don't think an behaviour has changed, and I think we should just go with the current one since the few versions we've tried are all functionally equivalent.

File: src/rest.cpp
Comment (jnewbery) [reply src/rest.cpp:117]: @fjahr I think the branch is invalid because you're trying to move from the global `g_chainman`.

In general, the usage of algebraic sum types like std::optional or std::variant isn't for prettiness - it's to entirely remove the possibility of being in an invalid program state.

Comment (fjahr) [reply src/rest.cpp:117]: Ok, never mind, I guess since there is no optional reference in std this wasn't a good idea after all.

**Pull request merged**
